$OpenBSD$

Built from avcall-powerpc.c with struct-copying code taken from hppa.

Index: avcall/avcall-powerpc-linux-macro.S
--- avcall/avcall-powerpc-linux-macro.S.orig
+++ avcall/avcall-powerpc-linux-macro.S
@@ -85,7 +85,7 @@ avcall_call:
 	cmpwi 0,9,1
 	beq- 0,.L34
 	cmpwi 0,9,0
-	beq- 0,.L84
+	beq- 0,.L78
 	cmpwi 0,9,2
 	beq- 0,.L77
 	cmpwi 0,9,3
@@ -97,24 +97,20 @@ avcall_call:
 	cmpwi 0,9,6
 	beq- 0,.L76
 	cmpwi 0,9,7
-	beq- 0,.L84
+	beq- 0,.L78
 	cmpwi 0,9,8
-	beq- 0,.L84
+	beq- 0,.L78
 	cmpwi 0,9,9
-	beq- 0,.L84
+	beq- 0,.L78
 	cmpwi 0,9,10
-	beq- 0,.L84
+	beq- 0,.L78
 	addi 0,9,-11
 	cmplwi 0,0,1
-	ble- 0,.L83
-	cmpwi 0,9,13
-	beq- 0,.L85
-	cmpwi 0,9,14
-	beq- 0,.L86
-	cmpwi 0,9,15
-	beq- 0,.L84
-	cmpwi 0,9,16
-	beq- 0,.L87
+	bgt- 0,.L55
+	lwz 9,8(30)
+	stw 4,4(9)
+.L75:
+	stw 3,0(9)
 .L34:
 	lwz 11,0(1)
 	li 3,0
@@ -124,44 +120,142 @@ avcall_call:
 	mtlr 0
 	mr 1,11
 	blr
-.L87:
+.L55:
+	cmpwi 0,9,13
+	beq- 0,.L79
+	cmpwi 0,9,14
+	beq- 0,.L80
+	cmpwi 0,9,15
+	beq- 0,.L78
+	cmpwi 0,9,16
+	bne+ 0,.L34
 	lwz 0,0(30)
 	andi. 9,0,512
 	beq- 0,.L34
-	lwz 0,16(30)
-	cmpwi 0,0,1
-	beq- 0,.L77
-	cmpwi 0,0,2
-	beq- 0,.L76
-	cmpwi 0,0,4
-	beq- 0,.L84
-	cmpwi 0,0,8
-	bne+ 0,.L34
-.L83:
-	lwz 9,8(30)
-	stw 4,4(9)
-.L75:
-	stw 3,0(9)
+	lwz 8,16(30)
+	addi 0,8,-1
+	cmplwi 0,0,7
+	bgt- 0,.L34
+	lwz 0,8(30)
+	cmplwi 0,8,4
+	rlwinm 9,0,0,30,31
+	rlwinm 5,0,0,0,29
+	add 8,9,8
+	bgt- 0,.L66
+	cmplwi 0,8,4
+	bgt- 0,.L67
+	slwi 0,9,3
+	slwi 10,8,3
+	subfic 0,0,31
+	li 11,2
+	lwz 8,0(5)
+	subfic 10,10,32
+	slw 11,11,0
+	slw 9,3,10
+	li 0,1
+	xor 9,8,9
+	slw 0,0,10
+	subfc 11,0,11
+	and 9,9,11
+	xor 8,8,9
+	stw 8,0(5)
 	b .L34
-.L84:
+.L67:
+	slwi 8,8,3
+	slwi 9,9,3
+	lwz 6,0(5)
+	subfic 9,9,31
+	lwz 7,4(5)
+	addi 0,8,-32
+	li 11,2
+	subfic 8,8,64
+	slw 11,11,9
+	slw 10,3,8
+	sraw 0,3,0
+	li 9,-1
+	xor 0,6,0
+	xor 10,7,10
+	addi 11,11,-1
+	slw 9,9,8
+	and 0,0,11
+	and 10,10,9
+	xor 6,6,0
+	xor 7,7,10
+	stw 7,4(5)
+	stw 6,0(5)
+	b .L34
+.L66:
+	slwi 0,9,3
+	cmplwi 0,8,8
+	subfic 0,0,31
+	li 9,2
+	slw 9,9,0
+	addi 12,9,-1
+	bgt- 0,.L70
+	slwi 9,8,2
+	slwi 10,8,3
+	addi 9,9,-16
+	subfic 10,10,64
+	sraw 11,4,9
+	lwz 8,0(5)
+	lwz 7,4(5)
+	sraw 11,11,9
+	slw 0,3,10
+	slw 4,4,10
+	or 0,0,11
+	li 9,-1
+	xor 0,8,0
+	xor 4,7,4
+	slw 9,9,10
+	and 0,0,12
+	and 4,4,9
+	xor 8,8,0
+	xor 7,7,4
+	stw 8,0(5)
+	stw 7,4(5)
+	b .L34
+.L70:
+	slwi 9,8,3
+	lwz 7,0(5)
+	addi 6,9,-64
+	lwz 8,8(5)
+	subfic 9,9,96
+	sraw 11,3,6
+	slw 10,4,9
+	li 0,-1
+	slw 0,0,9
+	xor 11,7,11
+	xor 10,8,10
+	and 11,11,12
+	and 10,10,0
+	slw 9,3,9
+	sraw 4,4,6
+	xor 7,7,11
+	or 9,9,4
+	xor 8,8,10
+	stw 8,8(5)
+	stw 7,0(5)
+	stw 9,4(5)
+	b .L34
+.L78:
 	lwz 9,8(30)
 	b .L75
-.L76:
+.L80:
 	lwz 9,8(30)
-	sth 3,0(9)
+	stfd 1,0(9)
 	b .L34
-.L77:
+.L79:
+	frsp 0,1
 	lwz 9,8(30)
-	stb 3,0(9)
+	stfs 0,0(9)
 	b .L34
-.L86:
+.L76:
 	lwz 9,8(30)
-	stfd 1,0(9)
+	sth 3,0(9)
 	b .L34
-.L85:
-	frsp 0,1
+.L77:
 	lwz 9,8(30)
-	stfs 0,0(9)
+	stb 3,0(9)
 	b .L34
 	.size	avcall_call, .-avcall_call
 #if defined __linux__ || defined __FreeBSD__ || defined __FreeBSD_kernel__ || defined __DragonFly__
